Feature: purchase orders

jniles opened the pull request:
This pull request adds a module for creating Purchase Orders.  The module is available at `/purchases/new`.

**Screen Shots**
![mainscreen](https://cloud.githubusercontent.com/assets/896472/16409876/aa74daa6-3d17-11e6-965c-b3f0a4d98e74.png)
_Fig 1: The main purchase order creation interface_.

![selectinginventoryitems](https://cloud.githubusercontent.com/assets/896472/16409884/b7934dd0-3d17-11e6-9b3b-62277963b792.png)
_Fig 2: Selecting Inventory Items_

The module does the very basic functionality for purchase order creation.
1. Allows an order to be created for any supplier
2. Performs basic form validation
3. Allows any inventory item to be selected only once.
4. Generates a PDF Purchase Receipt from the server.

There is more work to do:
1. Ensure that caching works correctly, and the rules should be written down.
2. Make the receipt look nice.
3. Decide how to calculate shipping charges, discounts, and extra fees.

Closes #523.

**PrototypeApiService Changes**

This PR also addresses some previous concerns with the `PrototypeApiService`.  I've written tests for every PrototypeApiService method, and started the best practice of requiring the service as `Api`.  The methods are then accessible, both at public methods on the `Api` function object itself, and as prototype methods.  It looks like this:

``` js
SomeService.$inject = [ 'PrototypeApiService' ];
function SomeService(Api) {
  var service = new Api('/some/url/');

  // service.read() is already written for us!  We'll find it in service.prototype.read() automagically!

  service.create = function create(data) {

     // We can do any kind of data manipulation we want before sending to the server.
     // It will be done in the function body, and then we can call the prototype create()
     // method to write data to the server.
     // for example, a console log.
     console.log('I will be fired before the regular create state');

     // call the PrototypeApiService's create method with the new or original arguments.
     return Api.create.call(this, data);
  }

  return service;
}
```

Closes #457.

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

jniles committed:
feat(purchase): complete purchases API

This commit implements the purchases API described in #512.  The schema
is updated to reflect new concerns and to remove the boolean states from
the database.  The integration tests are updated appropriately.

Closes #512.

jniles committed:
refactor(server): rm/mv unused routes

This commit completely removes some of the deprecated inventory and
purchasing server routes.  In the process, `categorised/` has been
removed.  The server is now much cleaner in terms of file structure,
allowing devs to iterate on the stucture they want without having all
the old modules getting in the way.

This commit specifically removes the `donor`s and `donation`s from the
database and server architecture.

jniles committed:
refactor(donors): remove donors client-side code

This commit removes the donors client-side code and tests.  It also
removes the donors paths in the test data.

Closes #480.

jniles committed:
refactor(server): remove data and parser

This commit removes the terrible data.js and parser.js from 1.X days.
It assures that both the client and server continue operating normally
without them.

greenkeeperio-bot committed:
chore(package): update stream-to-promise to version 2.1.1

https://greenkeeper.io/

jniles committed:
refactor(patient reg): finish 2.x review criteria (#511)

This commit cleans up the Patient Registrations page, attempting to
finalize the 2.x criteria for patient registrations and complete them
once and for all.

Significant Changes:
 1. The date input has been changed to a `uib-datepicker-popup`.
 Unfortunately, the datepicker doesn't actually validate the input, only
 the datepicker popup's calendar.  See
 https://github.com/angular-ui/bootstrap/issues/4664 for more
 information.  Therefore, I wrote a custom date `$validator`,
 `bhDateValidator` to ensure that dates are in the valid range.

 2. The patient registration pages uses `bhSubmit` to regulate form
 submission.

 3. Every single unused `id` was removed from the HTML inputs.

This commit should be reviewed to see if it may close the Patient
Registration review issue #205.

jniles committed:
deps(chart): update to chart-1.0 beta

jniles committed:
feat(purchases): client-side PO creation

This commit adds the purchase order client-side creation interface to
the 2.x client.  At the moment, form validation works, items can be
added to the grid, and suppliers selected from the database.

**Breaking Change**
The purchase order table now stores supplier_uuid instead of
creditor_uuid.  This intuitively makes more sense and makes easier
processing.  However, it required several purchase and supplier tests to
be updated, and therefore I am marking this as breaking.

jniles committed:
chore(services): improve & test API service

This commit improves the prototype API service with a cleaner API for
creating new services that inherit from it.  The new syntax is this:
```js
var service = new PrototypeApiService('url/');
// you can now call service.create(), service.update(), etc.
```

Additionally, the service has a new method `search()` to standardize the
way we send search queries.

Finally, the Prototype API service is now well tested with karma unit
tests.

Closes #421.

jniles committed:
Merge branch 'master' into feature-purchase-orders

Conflicts:
	bower.json
	server/config/routes.js
	server/controllers/finance/purchases.js
	server/models/schema.sql
	server/test/api/purchase.js

jniles committed:
feat(purchase): purchase client creates orders

This commit hooks up the client services with the database to ensure
that purchase orders can be posted to the server.  In order to
accomplish this, I changed the schema slightly to account for the
new `payment_method` field.  This field might need to be an ID instead
of a text field, but this works for a first implementation until we
decide upon the value of the method.

Next, a purchase receipt should be implemented followed by end to end
tests.

jniles committed:
feat(purchase): implement purchase receipt

This commit implements purchase order receipts.  It moves around some of
the inventory reports and receipts to give the inventory a well-defined
structure.  The purchase receipt has been put in `/inventory/`.  This
can change with discussion.

After this commit, only end to end tests remain.

jniles committed:
test(purchase): add e2e tests for purchase orders

This commit adds tests for the purchase orders client interface, heavily
inspired by those for patient invoicing.  It also rearranges the e2e
test directory so that all the patient tests are together, rather than
spread far apart.

jniles committed:
Merge branch 'master' into feature-purchase-orders

Conflicts:
	client/src/js/app.js

jniles committed:
fix(test): fix invoice tests

This commit fixes the expect()s from the invoice registry tests, given
that the order of testing has changed.  Instead of two initial invoices,
it should expect 4 and expect those filters to be okay as well.

jniles committed:
Merge remote-tracking branch 'upstream/master' into feature-purchase-orders

sfount commented on client/src/js/app.js:
```diff
@@ -373,6 +373,15 @@ function bhimaConfig($stateProvider, $urlRouterProvider, $urlMatcherFactoryProvi
+
+
+  .state('purchasesCreate', {
+    url : '/purchases/new',
```
We should formalise what we use for create paths - do you think we should use the `new` or `create` keyword?
@jniles @IMA-WorldHealth/local-contributors

jniles replied:
I'll switch mine to `/create`.  Create sounds better to me!

jniles committed:
Merge branch 'master' into feature-purchase-orders

Conflicts:
	client/src/i18n/en.json
	client/src/js/app.js
	client/src/js/services/receipts/ReceiptModal.js
	client/src/js/services/receipts/ReceiptService.js

sfount commented on client/src/js/services/DepotService.js:
```diff
@@ -9,15 +9,8 @@ DepotService.$inject = ['PrototypeApiService'];
-  // the url target
-  service.url = '/depots/';
+function DepotService(Api) {
+  var service = new Api('/depots/');
```
This reads well to me!

sfount commented on client/src/js/services/PurchaseOrderService.js:
```diff
@@ -1,53 +1,50 @@
+    });
+
+    // automatically set the currency_id to the enterprise currency
+    data.currency_id = Session.enterprise.currency_id;
```
Should this case be treated the same as https://github.com/IMA-WorldHealth/bhima-2.X/pull/538#discussion_r68750620?

jniles replied:
Yup!  I'll change it.

sfount commented on client/src/js/services/receipts/modal/receiptModal.tmpl.html:
```diff
@@ -20,7 +20,7 @@
 </div>
 
-<!-- Tempoarary Loading State -->
+<!-- Temporary Loading State -->
```
😆

jniles committed:
Merge branch 'master' into feature-purchase-orders

Conflicts:
	bower.json
	client/src/js/services/PrototypeApiService.js
	client/src/js/services/VoucherService.js

sfount commented on client/src/partials/purchases/create/PurchaseOrderForm.js:
```diff
@@ -0,0 +1,311 @@
+    var totals = this.totals;
+
+    // loop through the items summing them into a total
+    totals.rows = this.store.data.reduce(function (value, unit) {
```
It would probably be good to consistently either use the `var order = this;` alias to always use `this`, mixing the two can be a bit confusing reading through the code. 

The functionality will remain exactly the same but I think it would be good use `order` after referencing it here.

jniles replied:
Noted.  I'll make the changes.

sfount commented on client/src/partials/purchases/create/PurchaseOrderForm.js:
```diff
@@ -0,0 +1,311 @@
+    var order = this;
+
+    // copy the data so that forEach() doesn't get confused.
+    var cp = angular.copy(this.store.data);
```
[comment 6877513](https://github.com/IMA-WorldHealth/bhima-2.X/pull/537/files#r68777513)

jniles replied:
All this would be simple if we could use fat arrows on the client... :8ball:

sfount commented on client/src/partials/purchases/create/create.html:
```diff
@@ -0,0 +1,207 @@
+      <div class="col-xs-12">
+        <div class="panel panel-default">
+          <div class="panel-body">
+            <form name="PurchaseOrderForm" autocomplete="off" autocapitalize="off" autocorrect="off" novalidate>
```
If this was re-based it could use your [bh-form-defaults](https://github.com/IMA-WorldHealth/bhima-2.X/pull/542)

sfount commented on client/src/partials/purchases/create/create.html:
```diff
@@ -0,0 +1,207 @@
+                  </div>
+                </div>
+
+                <div class="form-group">
```
There is no `'has-error'` class here - will these show up highlighted if this is an invalid option in the form?

jniles replied:
The way it is currently set up... it's not possible to have an invalid option.  It defaults `on-delivery` in the controller and then it is a radio button, so there is no way to "uncheck" it, so to speak.

jniles replied:
Unless you can figure a way to make this invalid, that is ;)

sfount commented on client/src/partials/purchases/create/create.html:
```diff
@@ -0,0 +1,207 @@
+        <div class="btn-group pull-right" role="group">
+          <button
+            class="btn btn-default"
+            ng-class="{'btn-primary' : PurchaseCtrl.order.hasCacheAvailable() && PurchaseCtrl.order.details.supplier_uuid }"
```
This could be a cool design pattern (changing class on availability) - it would be good to think about how we could make it uniform across the application! These kind of design decisions will have to be formalised soon before the first 2.x stable.

jniles replied:
I would be really interested in exploring changing the recycle icon to a nice green text when cache is available.  I think it would look really nice.

jniles committed:
Merge branch 'master' into feature-purchase-orders

Conflicts:
	client/src/partials/templates/bhFindPatient.tmpl.html

sfount commented on client/src/partials/purchases/create/create.html:
```diff
@@ -0,0 +1,207 @@
+      </div>
+    </div>
+
+    <div id="purchase-order-grid" ui-grid="PurchaseCtrl.gridOptions" style="height : 230px; width : 100%;" ui-grid-auto-resize ui-grid-cellNav>
```
This would be a good candidate for `bhConstant` which landed in https://github.com/IMA-WorldHealth/bhima-2.X/pull/538 (After this pull request was submitted so entirely optional!) 

Something like: 

``` html
<div style="height : {{::PurchaseCtrl.Constants.INVOICE.GRID_HEIGHT}} />
```

(This could then be applied to patient invoices and any additional modules that use a similar format)

sfount commented on client/src/partials/purchases/create/create.js:
```diff
@@ -0,0 +1,126 @@
+  }
+
+  // fired whenever an input in the grid is changed.
+  function handleChange() {
```
Cool idea - reducing the number of times we have to digest and validate to only when it's required will improve performance a lot.

sfount commented on client/src/partials/purchases/create/templates/actions.tmpl.html:
```diff
@@ -0,0 +1,3 @@
+<div style="padding : 5px;" class="text-center">
```
If this could use 

``` html
class="ui-grid-cell-contents"
```

This provides the padding and assigns overflow rules (overflow : ellipses) although this looks like it's just a single icon.

I think it would be good to generally use this class on most cell templates.

sfount commented on client/src/partials/purchases/create/templates/amount.tmpl.html:
```diff
@@ -0,0 +1,3 @@
+<div style="padding : 5px; text-align : right;">
```
[ui-grid-cell-contents](https://github.com/IMA-WorldHealth/bhima-2.X/pull/537/files#r68903437)

sfount commented on client/src/partials/purchases/create/templates/code.tmpl.html:
```diff
@@ -0,0 +1,15 @@
+<div style="padding : 5px;">
```
I'm not sure about this one using the default cell class with the input and expected padding - if you could verify if this will work with the class and if not there is no issue here.

sfount commented on client/test/e2e/purchases/purchases.spec.js:
```diff
@@ -0,0 +1,126 @@
+    components.dateEditor.set(new Date('2016-03-03'));
+
+    // set the 'other' delivery method parameter
+    $('#other').click();
```
I think this is relevant to this discussion 
https://github.com/IMA-WorldHealth/bhima-2.X/pull/531#discussion_r68577726

Just for reference in the future!

jniles committed:
Merge branch 'master' into feature-purchase-orders

Conflicts:
	client/test/e2e/patient/invoice/registry.page.js
	client/test/e2e/patient_invoice/invoiceRegsitry.spec.js

jniles committed:
fix(purchase): implement feedback on PR

This commit implements the feedback from @sfount.
 1. bhConstants is used to set the grid height at the initial module
 load.
 2. The `purchases/new` path has been renamed to `purchases/create`.
 3. The relevant `ui-grid` cells leverage `ui-grid-cell-contents` to
 ensure proper overflows, padding, and centering on small screens.
 4. Resolved mixed aliases for `this` in services.
 5. Ensure that the `currency_id` is set on the server, not the client.
